table/table_test.cc

```diff
@@ -2231,73 +2231,109 @@ TEST_P(BlockBasedTableTest, BadChecksumType) {
 }
 
 namespace {
-std::string TrailerAsString(const std::string& contents,
-                            CompressionType compression_type,
-                            ChecksumType checksum_type) {
-  std::array<char, kBlockTrailerSize> trailer;
-  BlockBasedTableBuilder::ComputeBlockTrailer(contents, compression_type,
-                                              checksum_type, &trailer);
-  return Slice(trailer.data(), trailer.size()).ToString(/*hex*/ true);
+std::string ChecksumAsString(const std::string& data,
+                             ChecksumType checksum_type) {
+  uint32_t v = ComputeBuiltinChecksum(checksum_type, data.data(), data.size());
+
+  // Verify consistency with other function
+  if (data.size() >= 1) {
+    EXPECT_EQ(v, ComputeBuiltinChecksumWithLastByte(
+                     checksum_type, data.data(), data.size() - 1, data.back()));
+  }
+  // Little endian as in file
+  std::array<char, 4> raw_bytes;
+  EncodeFixed32(raw_bytes.data(), v);
+  return Slice(raw_bytes.data(), raw_bytes.size()).ToString(/*hex*/ true);
+}
+
+std::string ChecksumAsString(std::string* data, char new_last_byte,
+                             ChecksumType checksum_type) {
+  data->back() = new_last_byte;
+  return ChecksumAsString(*data, checksum_type);
 }
 }  // namespace
 
 // Make sure that checksum values don't change in later versions, even if
-// consistent within current version. (Other tests check for consistency
-// between written checksums and read-time validation, so here we only
-// have to verify the writer side.)
+// consistent within current version.
 TEST_P(BlockBasedTableTest, ChecksumSchemas) {
-  std::string b1 = "This is a short block!";
+  std::string b0 = "x";
+  std::string b1 = "This is a short block!x";
   std::string b2;
   for (int i = 0; i < 100; ++i) {
     b2.append("This is a long block!");
   }
-  CompressionType ct1 = kNoCompression;
-  CompressionType ct2 = kSnappyCompression;
-  CompressionType ct3 = kZSTD;
+  b2.append("x");
+  // Trailing 'x' will be replaced by compression type
+
+  std::string empty;
+
+  char ct1 = kNoCompression;
+  char ct2 = kSnappyCompression;
+  char ct3 = kZSTD;
 
   // Note: first byte of trailer is compression type, last 4 are checksum
 
   for (ChecksumType t : GetSupportedChecksums()) {
     switch (t) {
       case kNoChecksum:
-        EXPECT_EQ(TrailerAsString(b1, ct1, t), "0000000000");
-        EXPECT_EQ(TrailerAsString(b1, ct2, t), "0100000000");
-        EXPECT_EQ(TrailerAsString(b1, ct3, t), "0700000000");
-        EXPECT_EQ(TrailerAsString(b2, ct1, t), "0000000000");
-        EXPECT_EQ(TrailerAsString(b2, ct2, t), "0100000000");
-        EXPECT_EQ(TrailerAsString(b2, ct3, t), "0700000000");
+        EXPECT_EQ(ChecksumAsString(empty, t), "00000000");
```

Comment:
If the expected checksum value is changing, doesn't that mean the test is now broken?  I might be misreading the test but it sounds like those constants being compared to should never be changed or compatibility will fail.

Reply:
Sorry it's hard to see in the diff
![Screen Shot 2021-11-02 at 3 45 46 PM](https://user-images.githubusercontent.com/7980268/139962300-efc5e172-c3f7-4767-b1a0-9179b7200668.png)

Reply:
Aren't some of the later values actually different though?  

kCRC32c:  b1, ct2, t -- Was "012F9B0A57", now "2F9B0A57";

Some of the values appear to be different in the first non-zero character?

Reply:
The leading byte for "trailer" is the compression type. ChecksumAsString doesn't include this.

Reply:
Thanks for the explanation.  Not obvious!